lib/ts/recipe/multifactorauth/recipe.ts

```diff
@@ -0,0 +1,508 @@
+                userContext,
+            });
+
+            if (canCreatePrimary.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+                // Race condition since we just checked that it was not a primary user
+                this.querier.invalidateCoreCallCache(userContext);
+                return {
+                    status: "RECURSE_FOR_RACE",
+                };
+            }
+
+            if (canCreatePrimary.status === "ACCOUNT_INFO_ALREADY_ASSOCIATED_WITH_ANOTHER_PRIMARY_USER_ID_ERROR") {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_011)",
+                };
+            }
+        }
+
+        // Check if there if the linking with session user going to fail and avoid user creation here
+        const users = await listUsersByAccountInfo(
+            tenantId,
+            { email: accountInfo.email, phoneNumber: accountInfo.phoneNumber },
+            true,
+            userContext
+        );
+        for (const user of users) {
+            if (user.isPrimaryUser && user.id !== sessionUser.id) {
+                return {
+                    status: "VALIDATION_ERROR",
+                    reason: "Cannot complete MFA because of security reasons. Please contact support. (ERR_CODE_012)",
+                };
+            }
+        }
+
+        return {
+            status: "OK",
+        };
+    };
+
+    linkAccountsForFactorSetup = async (
+        sessionUser: User,
+        factorUserRecipeUserId: RecipeUserId,
+        userContext: UserContext
+    ): Promise<{ status: "OK" | "RECURSE_FOR_RACE" }> => {
+        // if we are here, it means that all the validations passed in the first place. So any error
+        // in this function must result in retry from the validation.
+
+        // At this point, we have the recipe user for the new factor created. This means that
+        // when retrying for passwordless / thirdparty signInUp, where we check for existing user,
+        // we are going to find the user with the account info, technically converting this from
+        // sign up to a sign in operation. We need this behaviour to make the API repeatable.
+
+        // For emailpassword sign up, when we retry, the return point would be from the validation.
+
+        if (!sessionUser.isPrimaryUser) {
+            const createPrimaryRes = await AccountLinkingRecipe.getInstance().recipeInterfaceImpl.createPrimaryUser({
+                recipeUserId: new RecipeUserId(sessionUser.id),
+                userContext,
+            });
+            if (createPrimaryRes.status === "RECIPE_USER_ID_ALREADY_LINKED_WITH_PRIMARY_USER_ID_ERROR") {
+                this.querier.invalidateCoreCallCache(userContext);
+                return {
+                    status: "RECURSE_FOR_RACE",
+                };
```

Comment:
this is not going to solve the issue cause the session boolean will not change?!

Reply:
Sorry.. invalid comment on my side. The code has sessionUser.isPrimary which is not checking the session.. so this may be valid. But double check.

Reply:
Yes, have updated that case to return TRY_REFRESH_TOKEN